contract/services.py

```diff
@@ -610,51 +613,70 @@ def __create_contribution_from_policy(self, ccpd, policies):
             ccpd.save(username=self.user.username)
             return [ccpd_new, ccpd]
 
-    def __get_policy(self, insuree, date_valid_from, date_valid_to, product_id):
+    def __get_policy(self, insuree, date_valid_from, date_valid_to, product):
         from core import datetime
         policy_output = []
         # get all policies related to the product and insuree
-        policies = Policy.objects.filter(product__id=product_id).filter(family__head_insuree=insuree)
-        # get covered policy
-        if len(list(policies)) > 0:
-            policies_covered = list(policies.filter(
-                Q(start_date__lte=date_valid_to, expiry_date__gte=date_valid_from)
-            ).order_by('start_date'))
+        policies = Policy.get_queryset(Policy.objects.filter(product=product)\
+                    .filter(family__head_insuree=insuree)\
+                    .filter(start_date__lte=date_valid_to, expiry_date__gte=date_valid_from))
+        # get covered policy, use count to run a COUNT query
+        if policies.count() > 0:
+            policies_covered = list(policies.order_by('start_date'))
         else:
             policies_covered = []
-
+        missing_coverage = []
+        # make sure the policies covers the contract : 
         last_date_covered = date_valid_from
+        # get the start date of the new contract by updating last_date_covered to the policy.stop_date
         while last_date_covered < date_valid_to and len(policies_covered) > 0:
             cur_policy = policies_covered.pop()
             # to check if it does take the first
-            if cur_policy.start_date < date_valid_from:
+            if cur_policy.start_date <= last_date_covered:
                 # Really unlikely we might create a policy that stop at curPolicy.startDate
                 # (start at curPolicy.startDate - product length) and add it to policy_output
                 last_date_covered = cur_policy.expiry_date
                 policy_output.append(cur_policy)
+            elif cur_policy.expiry_date <= date_valid_to:
+                missing_coverage.append({'start': cur_policy.start_date, 'stop': last_date_covered})
+                last_date_covered = cur_policy.expiry_date
+                policy_output.append(cur_policy)
+
+        for data in missing_coverage:
+            policy_created = create_contract_details_policies(insuree, product, data['start'], data['stop'])
+            if policy_created is not None and len(policy_created)>0:
+                policy_output += policy_created
 
         # now we create new policy
         while last_date_covered < date_valid_to:
```

Comment:
@delcroip In this line I received infinitive loop. This condition is still True.

Reply:
just to add 'last_expired_date' in return. I will push this and retest.